Package system on libgit2

wildart opened the pull request:
Package system switch to `libgit2` promises faster and resource efficient package management.

This work originated in #11143.

**Pending:**
- [x] next libgit2 release, [tracking issue here](https://github.com/libgit2/libgit2/issues/3135) (will [provide support](https://github.com/libgit2/libgit2/pull/2986) for HTTPS on Windows sans OpenSSL)
- [x] [homebrew update for libgit2 0.23](https://github.com/Homebrew/homebrew/pull/41578)

wildart committed:
changing `installed` based to support `libgit2`

wildart committed:
merged installed & dependents into `Read`, moved `libgit2` calls into libgit.jl

wildart committed:
removed `Git` module calls from `Read`
`Repo` object auto-frees itself

wildart committed:
changed string type

wildart committed:
renamed to 'libgit2'
moved parts to `libgit2` dir
converted github.jl
refactored repository and configuration types (as in LibGit2 pkg)

wildart committed:
fixed to AbstractString

wildart committed:
added libgit2 error handling

wildart committed:
write to repo configuration value with any accepted type

wildart committed:
added `clone` & `fetch`, updated 'Dir' and `Cache`

wildart committed:
switching to `Ref` type

wildart committed:
changed incorrect types & syntax

wildart committed:
working on `Generate`
added multiples types, refactored files in `libgit2`
it builds!!!

wildart committed:
fixed changed api

IainNZ commented:
This is heroic stuff

wildart committed:
construct type pragmatically
get rid off finalizers & do manual memory management everywhere
added signature, commit, remote

wildart committed:
added `with_libgit2` helper function
better memory resource handling
`finalize` instead of `free`

wildart committed:
package generator works!!!
added `index` & `strarray` related functions

nalimilan commented on base/pkg/libgit2/const.jl:
```diff
@@ -0,0 +1,210 @@
+module GitConst
+
+    const OBJ_ANY    = Cint(-2)
```
Wouldn't it make sense to use enums here now that they are supported?

nalimilan commented on base/pkg/libgit2/error.jl:
```diff
@@ -0,0 +1,107 @@
+    const ITEROVER        = Cint(-31)
+end
+
+const git_error_code = Dict{Int,Symbol}(
```
Here in particular, if you use an enum in the block above you should be able to do things like `GitError[-03]` and get `notfound` without defining the dict manually.

wildart replied:
Yeah, good idea. Thanks.

jakebolewski commented on base/pkg/libgit2/oid.jl:
```diff
@@ -0,0 +1,116 @@
+#   ...
+#   id20::UInt8
+# end
+@eval begin
```
The Oid generated code should now use an immutable type that wraps an ntuple of length 20

jakebolewski commented on base/pkg/libgit2/signature.jl:
```diff
@@ -0,0 +1,37 @@
+                 (Ptr{Ptr{SignatureStruct}}, Ptr{Uint8}, Ptr{Uint8}, Cint, Cint),
+                 sig_ptr_ptr, sig.name, sig.email, sig.time, sig.time_offset)
+    return GitSignature(sig_ptr_ptr[])
+end
```
All the git signature stuff was terribly memory unsafe.  It really should be rewritten.

wildart replied:
This part should be fine. Memory allocated in `git_signature_default` is cleared in `finalize` (Only wrapper types have `finalize` functions that clear wrapped object to which wrapper holds the pointer).  Nothing more is allocated. I use `GitSignature` as a wrapper around pointer to `SignatureStruct` and `Signature` only reads data from `SignatureStruct`.

jakebolewski replied:
Ah, I see you changed this quite a bit.  Yes, making copies everywhere is the right way to go.

jakebolewski commented on base/pkg/libgit2/strarray.jl:
```diff
@@ -0,0 +1,29 @@
+    sa_ptr = Ref(sa)
+    ccall((:git_strarray_free, :libgit2), Void, (Ptr{StrArrayStruct},), sa_ptr)
+    return sa_ptr[]
+end
```
Same here.  I don't know if there have been any updates in the past to libgit year that would make this less hacky.

jakebolewski commented on base/pkg/libgit2/error.jl:
```diff
@@ -0,0 +1,107 @@
+end
+
+immutable GitError{Class, Code}
+    msg::AbstractString
```
This should be a <: of Exception.

jakebolewski commented on base/pkg/libgit2/reference.jl:
```diff
@@ -0,0 +1,25 @@
+    typ == 1 && return ""
+    rname = ccall((:git_reference_symbolic_target, :libgit2), Ptr{UInt8}, (Ptr{Void},), ref.ptr)
+    rname == C_NULL && return ""
+    return bytestring(rname)
```
I would double check here if libgit returns a string that you are supposed to free, or just a pointer that is managed by libgit.  Otherwise this will be a memory leak.

wildart replied:
[Returned strings are part of reference](https://github.com/libgit2/libgit2/blob/v0.22.2/src/refs.c#L1268), so they handled when reference is freed. It would be better to copy output to a new string.

jakebolewski replied:
Ok, `bytestring` returns a copy so this is fine.  Always worth double checking, but I think it got a lot more consistent when they (more recently) introduced the string array buffer type in libgit2.

jakebolewski commented on base/pkg/libgit2/types.jl:
```diff
@@ -0,0 +1,193 @@
+    end
+
+    @eval Base.isempty(obj::$typ) = (obj.ptr == C_NULL)
+end
```
I don't think this really makes sense

jakebolewski replied:
this is is empty definition I mean.

jakebolewski replied:
Some of the above objects should also be <: of GitObject.

wildart replied:
Maybe some `AbstractGitObject` supertype for all wrappers.
```julia
abstract AbstractGitObject
type GitRepo <: AbstractGitObject
abstract GitObject <: AbstractGitObject
type GitTree <: GitObject
Base.isempty(obj::AbstractGitObject) = (obj.ptr == C_NULL)
```
So, there would be only one `isempty` method.

wildart committed:
added default value to GitConfig 'get' to prevent exceptions

wildart committed:
added anyobject type and `checkout` functions

wildart committed:
no more `git` in Write

wildart committed:
fixed `revparse`, `isdirty` & errors in `Write`

wildart committed:
improved git types description

wildart committed:
transitioned Entry.add

wildart committed:
transitioned Entry.clone

vtjnash commented on base/pkg/libgit2/error.jl:
```diff
@@ -0,0 +1,107 @@
+    class::Cint
+end
+
+immutable GitError{Class, Code}
```
this class should not be parameterized, as it is terrible for memory usage to generate this many unique types.

jakebolewski replied:
It can generate at most 27, but in hopefully in real code you would only generate a couple of unique types as you don't exercise all possible error paths.  I don't see how this is any worse than making an enum of the error values.

vtjnash replied:
class*code = 27*15 = 405?

`@enum` creates values not types for each enumeration.

wildart replied:
I rewrote error handling with `@enum`.

wildart committed:
GitError is Exception subclass

wildart committed:
fixed `need_update`, `head_oid`, `revparse`, `CHECKOUT_NONE`, `CheckoutOptionsStruct`
added index: `read!` & repo: `checkout`, `new_branch`

wildart committed:
transitioned Entry: `free` & `pin`

wildart committed:
refactored error handling

wildart committed:
error fix: nonexistent directory case, cannot create empty wrapper

wildart committed:
made options structures immutable with convinient constructors

wildart committed:
rewrote Write module: changed resource management and commit checkout.
some minor fixes and refactoring

wildart committed:
added proper `checkout` & `branch` functions
errors fixed & refactored code

wildart committed:
added tests for `libgit2`
added functions: `reset` & `checkout_head`
merged multiple `fetch` functions into one
fixed `restore` & `transact`

tkelman commented:
You're doing a kickass job in here, btw.

ViralBShah commented:
This is fantastic. I wonder the likelihood of having this go into 0.4.

hayd commented on base/pkg/cache.jl:
```diff
@@ -25,25 +25,41 @@ function mkcachedir()
+        try
+            LibGit2.GitRepo(cache) # open repo, free it at the end
+        catch err
+            rethrow(err)
```
Does this catch block do anything?

wildart replied:
Exactly. :+1:

nalimilan replied:
Funny answer for a question. :-p

hayd commented on base/pkg/cache.jl:
```diff
@@ -25,25 +25,41 @@ function mkcachedir()
+            LibGit2.clone(url, cache, bare = true, remote_cb = LibGit2.mirror_cb)
+        catch err
+            isdir(cache) && rm(cache, recursive=true)
+            error("Cannot clone $pkg from $url\n", err)
```
I wonder if there ought to be a PkgError type, rather than error (throughout).

wildart replied:
Usually, after these errors the package system interrupts and no exception handling is required. I guess  it should be done properly. I see your point.

tkelman replied:
+1 for PkgError, could help disambiguate exceptions if someone wants to script Pkg in clever ways

wildart commented:
@ViralBShah That was an idea, but it depends on 0.4 release date. At this point, there are few places left to convert. I didn't expect that `libgit2` would be such a present. I count for another week to finish rewrite.  After all, this thing would require some thorough testing.

wildart committed:
removed unnecessary `rethrow` calls
improved `branch!`
removed `dir` parameter from `transact`

tkelman commented:
We're still on -dev for a little while longer, but probably not too much. If you can get this fully working on all platforms in the next few weeks, then we should think about switching. I'll need a few days to do some surgery to take command-line git out of the Windows binaries (and similar but not quite as invasive for the Mac binaries), if we're going to be ambitious enough for a complete replacement.

We could possibly merge a conservative version of this for 0.4 that leaves both command-line git and libgit2 simultaneously around, and only uses libgit2 for the parts of Pkg where you feel it will work robustly.

wildart commented:
@tkelman I do not have access to OSX. Help appreciated.

tkelman commented:
Neither do I. I can help you tweak appveyor to run the pkg test when you think it's ready for that, for now it isn't building this PR since it's apparently un-mergeable, but that's probably for the best at the moment.

tkelman commented:
It'll probably still time out, but may as well uncomment this line https://github.com/JuliaLang/julia/blob/713e9e55f7cf5f53416ddb05d501b3a1d61288a9/.travis.yml#L4 to get this PR testing on OSX

wildart committed:
transitioned `checkout`, `free`, `update`, `register`
added merge functionality

wildart committed:
switched to https

wildart committed:
added `gitdir` (location of repo '.git') and `path` (repo location)

wildart committed:
import Base for `finalize`s
better warning information
added `libgit2` tags

wildart committed:
transitioned `tag` & `register`
code cleanup & refactoring
by default https url generated
added blob type

wildart committed:
fixed `get` with default for GitConfig
fixed `libgit2` tests

wildart committed:
fixed errors, set to run only pkg & libgit2 tests

tkelman commented:
There was some broken homebrew stuff on OSX Travis that I fixed, and it looks like this passes! Worth rebasing, and tweaking appveyor now you think?

Also good to add the new license header style on any newly added files

ViralBShah commented:
I am happy to test on OS X, and I am sure so will @StefanKarpinski

wildart committed:
added PkgError exception

wildart committed:
added licence line

wildart committed:
fixed GitHub saved token retrieval & PkgError visibility

wildart committed:
added `push`, transitioned `pull_request`

stevengj commented:
It is great to see this work happening.  Do you have a sense of how much performance improvement we will get, particularly on Windows?  Will `Pkg.update` no longer require a coffee break?   Any benchmark numbers you have would be great to see.

wildart committed:
fixed import error

wildart commented:
As I understand, some of the Windows trouble relates to a slow `git` util performance. This patch should fix these problems. Memory usage will go down as well, several folds.
However, package dependency resolution system takes some time, currently it is a major source of slowdown. And as the METADATA repo will grow, updates will take more time as well.

wildart commented:
@tkelman Can you update OSX `libgit2` to 0.22.2? Just in case, even though test were passed.

tkelman commented:
It'll need to be done at upstream homebrew https://github.com/Homebrew/homebrew/blob/master/Library/Formula/libgit2.rb

wildart commented:
Oh, I hope it wouldn't cause trouble.

wildart committed:
added `revcount` for counting differences

wildart committed:
fixed exceptions in tests

wildart commented:
@tkelman It looks like `libgit2` Windows dependency is not build with OpenSSL support.

tkelman commented:
That's correct. We don't have the mess of compiling openssl on Windows put into the build system. I had warned about this in a few previous issues. At the next release of libgit2 we should no longer need openssl, I cleaned up and got the PR merged so libgit2 on master can use the Windows native SSL functionality when compiled with MinGW. As a stopgap, we can get libgit2 with openssl support downloaded from opensuse into win-extras, like we do right now with libgcc dll's and expat and zlib.

StefanKarpinski commented:
This is great work, @wildart. Happy to test on OS X, of course.

stevengj commented on base/pkg/cache.jl:
```diff
@@ -28,27 +28,33 @@ function mkcachedir()
+            info("Updating cache of $pkg...")
+            LibGit2.fetch(repo)
+        end
+        filter(sha1->!LibGit2.iscommit(sha1, repo), sha1s)
```
This code seems to potentially run `LibGit2.iscommit` twice on all the `sha1s`.   Maybe first run `iscommits = [LibGit2.iscommit(sha1, repo) for sha1 in sha1s]` and then use `!all(iscommits)` and `sha1s[iscommits]`?

tkelman replied:
That sounds me a lot like the approach #11137 used...

stevengj commented on base/pkg/entry.jl:
```diff
@@ -285,21 +324,31 @@ function update(branch::AbstractString)
+        info("Pushing changes as branch $branch")
+        LibGit2.push(repo, fork, "$commit:refs/heads/$branch")
+        pr_url = "$(response["html_url"])/compare/$branch"
+        @osx? run(`open $pr_url`) : info("To create a pull-request, open:\n\n  $pr_url\n")
```
Probably we should have a function `openbrowser(url, message)`.   On Linux, we could use `xdg-open url`, and on [Windows we could use](http://stackoverflow.com/questions/3739327/launching-a-website-via-windows-commandline) `explorer url`.

one-more-minute replied:
c.f. [Blink.jl](https://github.com/one-more-minute/Blink.jl/blob/62408937f3a2682b934c7ad54111b754218739e3/src/content/config.jl#L15-L17)'s version

wildart replied:
Should browser be opened?

tkelman replied:
I don't think it has to. Usually when you navigate to the corresponding repo, github will notice and have a banner with an "open PR" button any time you have a recently-pushed branch in your fork.

stevengj commented on base/pkg/generate.jl:
```diff
@@ -2,25 +2,28 @@
-github_user() = readchomp(ignorestatus(`git config --global --get github.user`))
+copyright_year() =  string(Dates.year(Dates.today()))
+copyright_name(repo::GitRepo) =
+    with(GitConfig, repo) do cfg
```
I see this `with(GitConfig, repo) do cfg .... ` pattern a lot in the code.  Seems like it would be better to abstract it as a `getconfig(repo, key, default="")` function.

wildart replied:
It makes sense. :+1:

stevengj commented on base/pkg/libgit2.jl:
```diff
@@ -0,0 +1,438 @@
+    clone_opts_ref = Ref(clone_opts)
+    repo_ptr_ptr = Ref{Ptr{Void}}(C_NULL)
+    @check ccall((:git_clone, :libgit2), Cint,
+            (Ptr{Ptr{Void}}, Ptr{UInt8}, Ptr{UInt8}, Ref{CloneOptionsStruct}),
```
Here and elsewhere, you should use `Cstring`, not `Ptr{UInt8}`, when passing Julia strings as arguments to `ccall` where a NUL-terminated UTF-8 string is expected.

stevengj commented on base/pkg/libgit2.jl:
```diff
@@ -0,0 +1,438 @@
+        with(GitIndex, repo) do idx
+            content = readdir(path(repo))
+            if length(content) > 1
+                files = [utf8(bytestring(c))::UTF8String for c in content]
```
#11192 added a `utf8(c)` method, so you don't need `utf8(bytestring(c))`.

stevengj commented on base/pkg/libgit2/blob.jl:
```diff
@@ -0,0 +1,5 @@
+
+function content(blob::GitBlob)
+    return ccall((:git_blob_rawcontent, :libgit2), Ptr{Void}, (Ptr{Void},), blob.ptr)
+end
```
missing newline.  why a file with only one function?

wildart replied:
WIP, final refactoring is coming.

stevengj commented on base/pkg/libgit2/commit.jl:
```diff
@@ -0,0 +1,89 @@
+        finalize(comm_sig)
+    end
+    return commit_id
+end
```
newline would be nice

stevengj commented on base/pkg/libgit2/config.jl:
```diff
@@ -0,0 +1,78 @@
+    return GitConfig(cfg_ptr_ptr[])
+end
+
+function get{T}(::Type{T}, c::GitConfig, name::AbstractString)
```
These ifelse clauses are very non-Julian.  It would be better to have separate `get{T<:AbstractString}(::Type{T}, ...)`, `get(::Type{Bool}, ...)` etcetera methods.

stevengj commented on base/pkg/libgit2/config.jl:
```diff
@@ -0,0 +1,78 @@
+    return res
+end
+
+function set!{T}(c::GitConfig, name::AbstractString, value::T)
```
similar to `get` above: use method dispatch

wildart replied:
What do you mean?

hayd replied:
As in define for each type:

```
function set!(c::GitConfig, name::AbstractString, value::AbstractString)
    @check ccall((:git_config_set_string, :libgit2), Cint,
                 (Ptr{Void}, Ptr{UInt8}, Ptr{UInt8}), c.ptr, name, value)
end
function set!(c::GitConfig, name::AbstractString, value::Bool)
    bval = Int32(value)
    @check ccall((:git_config_set_bool, :libgit2), Cint,
                 (Ptr{Void}, Ptr{UInt8}, Cint), c.ptr, name, bval)
end
# etc
```
if you need to use T (as above) do:

```
function get{T<:AbstractString}(::Type{T}, ....
```

wildart replied:
I see. Thanks.

stevengj commented on base/pkg/libgit2/const.jl:
```diff
@@ -0,0 +1,224 @@
+    const OBJ_TAG    = Cint(4)
+
+    const SORT_NONE = Cint(0)
+    const SORT_TOPOLOGICAL = Cint(1) << Cint(0)
```
`Cint(1 << 0) ` etc. seems cleaner.

stevengj commented on base/pkg/libgit2/const.jl:
```diff
@@ -0,0 +1,224 @@
+    const DIFF_LINE_BINARY    = Cchar('B')
+
+    # index
+    const IDXENTRY_NAMEMASK   = (0x0fff)
```
extra parens

stevengj commented on base/pkg/libgit2/const.jl:
```diff
@@ -0,0 +1,224 @@
+    const INDEXCAP_IGNORE_CASE = Cuint(1)
+    const INDEXCAP_NO_FILEMODE = Cuint(2)
+    const INDEXCAP_NO_SYMLINKS = Cuint(4)
+    const INDEXCAP_FROM_OWNER  = ~(Cuint(0))
```
extra parens

stevengj commented on base/pkg/libgit2/const.jl:
```diff
@@ -0,0 +1,224 @@
+    const MERGE_FASTFORWARD_ONLY = Cint(2)
+
+    const MERGE_ANALYSIS_NONE = 0,
+    const MERGE_ANALYSIS_NORMAL = (1 << 0)
```
extra parens ... why are these constants `Int`, unlike all the others?

stevengj commented on base/pkg/libgit2/const.jl:
```diff
@@ -0,0 +1,224 @@
+    const RESET_SOFT  = Cint(1) # Move the head to the given commit
+    const RESET_MIXED = Cint(2) # SOFT plus reset index to the commit
+    const RESET_HARD  = Cint(3) # MIXED plus changes in working tree discarded
+end
```
newline

stevengj commented on base/pkg/libgit2/index.jl:
```diff
@@ -0,0 +1,118 @@
+    ie_ptr = ccall((:git_index_get_byindex, :libgit2), Ptr{Void},
+                  (Ptr{Void}, Csize_t), idx.ptr, i)
+    ie_ptr == C_NULL && return nothing
+    return unsafe_load(convert(Ptr{IndexEntry}, ie_ptr), 1)
```
You don't need the `1` argument to `unsafe_load`

stevengj commented on base/pkg/libgit2/oid.jl:
```diff
@@ -0,0 +1,96 @@
+function hex(id::Oid)
+    _hexstr = zeros(UInt8, OID_HEXSZ)
+    ccall((:git_oid_nfmt, :libgit2), Void,
+          (Ptr{UInt8}, Csize_t, Ptr{Oid}), Ref(_hexstr), OID_HEXSZ, Ref(id))
```
This should just be `_hexstr`, not `Ref(_hexstr)`.

Also, the third argument should be declared as `Ref{Oid}`, and then you just pass `id`, not `Ref(id)`.  Similarly for almost all of the other uses of `Ref` in the code.   You only need to explicitly construct a `Ref` object if you want to get data back through that argument (as in e.g. `git_oid_fromraw`).

stevengj commented on base/pkg/libgit2/reference.jl:
```diff
@@ -0,0 +1,154 @@
+        @check ccall((:git_reference_create, :libgit2), Cint,
+                 (Ptr{Ptr{Void}}, Ptr{Void}, Ptr{UInt8}, Ptr{Oid}, Cint, Ptr{SignatureStruct}, Ptr{UInt8}),
+                  ref_ptr_ptr, repo.ptr, refname, Ref(obj_oid), Cint(force),
+                  sig.ptr, isempty(msg) ? Ptr{UInt8}(C_NULL) : msg)
```
See above regarding `Cstring` here and elsewhere.  (`Ptr{UInt8}` is problematic because it will silently truncate any string that has an embedded NUL char: #10994).

wildart replied:
I find usage of the `Cstring` problematic because I cannot initialize it with `C_NULL` nor create an object of it. I have to do something like this, `convert(Cstring, Ptr{Uint8}(C_NULL))`, in order to get NULL value of `Cstring` type.

vtjnash replied:
I'm sure that could be fixed with an extra method on convert. 

Alternatively (like GtkNullContainer), we could define a CNullString type (and value) that implements this behavior.

stevengj replied:
I would just add a Cstring() constructor that makes a null

stevengj commented on base/pkg/libgit2/strarray.jl:
```diff
@@ -0,0 +1,25 @@
+    for i=1:count
+        len = length(strs[i])
+        #in_ptr  = convert(Ptr{Uint8}, bytestring(strs[i]))
+        in_ptr  = pointer(bytestring(strs[i]))
```
Use `unsafe_convert(Cstring, strs[i])` here, in order to throw an error if the strs contain NULL chars.   (Then you have to also wrap the whole thing in a `try ... finally` construct in order to avoid memory leaks if an exception is thrown.)

vtjnash replied:
This looks a lot like the existing Ref{Ptr{UInt8}}(strs) constructor

Sizeof(UInt8) is defined to be 1, so unnecessary to compute it

(From iphone)

stevengj commented on base/pkg/libgit2/strarray.jl:
```diff
@@ -0,0 +1,25 @@
+    end
+    return StrArrayStruct(strings, count)
+end
+StrArrayStruct{T<:AbstractString}(strs::Vector{T}) = StrArrayStruct(strs...)
```
Seems like this should be the other way around: your "main" constructor should be the `Vector{T}

stevengj commented on test/libgit2.jl:
```diff
@@ -17,3 +17,196 @@ function check_version()
+        @test arr[1] == p1
+        @test arr[2] == p2
+    finally
+        Pkg.LibGit2.finalize(sa2)
```
Just `finalize`, since `Base.finalize` was extended.

wildart replied:
Yes, I remember do that.

stevengj commented on base/pkg/libgit2/types.jl:
```diff
@@ -0,0 +1,380 @@
+                     remote_cb_payload)
+
+# git diff option struct
+immutable DiffOptionsStruct
```
Did you forget to define `Base.finalize` for this?  (The finalizer needs to finalize `pathspec`.)

wildart committed:
speedup `prefetch` call

wildart committed:
refactored GitConfig methods

wildart committed:
removed prefix from `finalize` calls

wildart committed:
newline at end of file

tkelman commented:
It looks like libgit2 will be making their next release pretty soon, ref https://github.com/libgit2/libgit2/issues/3135

Probably worth some quick tests against libgit2 master to see if there might be any issues we could report upstream in time.

wildart commented:
Problem is how fast Homebrew and OBS (WinRPM) can adopt it. They both are on 0.22.1 now.

tkelman commented:
I'm suggesting people try locally, before the next release, if at all possible. WinRPM is at 0.22.2, Homebrew should be just a matter of a very simple PR.

wildart committed:
No more `Git`!
updated `merge`, `fetch`, `rebase`, `push`, `diff` & many small fixes
general cleanup and refactoring

hayd commented on base/pkg/cache.jl:
```diff
@@ -28,27 +28,35 @@ function mkcachedir()
-    if !all(in_cache)
-        info("Updating cache of $pkg...")
-        Git.success(`remote update`, dir=cache) ||
-        error("couldn't update $cache using `git remote update`")
```
Potentially worth noting - IIUC - that this is an API change: this no longer errors if there is a problem with fetch (LibGit2.fetch just warns). New behaviour is better IMO.

wildart replied:
I did not intend to warn on `fetch` failure. I was a working version. I guess it is fine to leave it this way.

hayd commented on base/pkg/entry.jl:
```diff
@@ -11,6 +13,7 @@ macro recover(ex)
         try $(esc(ex))
         catch err
             show(err)
+            print('\n')
```
This seems random...

wildart replied:
`show` doesn't print new line, so I added it, otherwise an output on a screen is mangled.

hayd replied:
Would an alternative solution be to append '\n' to the error/show?

wildart replied:
`show` just prints string as it is. It doesn't process whitespace chars. Try `show("\n")`

stevengj replied:
why not `println(err)`?

tkelman commented on base/pkg/libgit2/types.jl:
```diff
@@ -0,0 +1,419 @@
+                        ancestor_label::Ptr{UInt8} =Ptr{UInt8}(0),
+                        our_label::Ptr{UInt8} = Ptr{UInt8}(0),
+                        their_label::Ptr{UInt8} = Ptr{UInt8}(0)
+)=CheckoutOptionsStruct(one(Cuint),
```
this constructor style is pretty messy, it looks like you could really use a default-values macro to make all of these much shorter

wildart committed:
partial fix for #11580

wildart committed:
adopting CString & refactoring

wildart committed:
added prototype of git repl

tkelman commented:
There was an 0.23.0-RC1 tag upstream. Hopefully the release will follow in a week or so. Would help to rebase this branch to prepare for that. We can upgrade the version on master as soon as there's a release tag, that will get it into the nightlies.

wildart committed:
added `clone` & `init` command

wildart committed:
refactoring commit message call

tkelman commented:
@wildart master is now using (a slightly patched) libgit2 v0.23.0, so https on windows should work now if you have time to rebase this

testurser committed:
rebased, moved `libgit2` to `base`, updated to 0.23 & tests

wildart committed:
fixed repl tests

testurser committed:
tuple-based Oid

testurser committed:
added `Oid` tests, modified `free` with tuple parameters,
updated `git_reference` calls to 0.23, fixed `branch!`, code cleanup

tkelman commented:
```
ERROR: LoadError: unlink: operation not permitted (EPERM)
 in unlink at fs.jl:102
while loading C:\projects\julia\test\libgit2.jl, in expression starting on line 112
```

You need to close open files when you're done with them.

wildart commented:
It relates to test fixer, I delete repository folder after all tests are done on it. I guess `rm` command does poor job on removing directories with links, [Libc manual: Deleting Files](http://www.gnu.org/software/libc/manual/html_node/Deleting-Files.html).

tkelman commented:
Windows doesn't use glibc. I'll bet you a beer that it's because you're trying to delete files for which you still have handles open.

wildart committed:
fixed: string value retrieval from config, resource management in tests

wildart commented:
@tkelman We are getting closer, win-x86_64 passed. What do you think of win-32 build error?  How can we speed up Homebrew update to 0.23?

tkelman commented:
> What do you think of win-32 build error?

Segfault in `git_filebuf_commit` - what does that function do?

> How can we speed up Homebrew update to 0.23?

https://github.com/Homebrew/homebrew/pull/41578 - though I might need help from someone who owns a mac if I the homebrew maintainers want me to do anything else

wildart commented:
`git_filebuf_commit` is deep inside. I never heard of it until now. [There is nothing special: flushing & closing file.](https://github.com/libgit2/libgit2/blob/v0.23.0/src/filebuf.c#L318). I'll try to change win-32 build to make a debug build of libgit2 for more information.

tkelman commented:
> I'll try to change win-32 build to make a debug build of libgit2 for more information.

How so? We don't build libgit2 from source on appveyor right now, though things could possibly be tweaked to do so.

wildart commented:
Can we use for now debug Windows versions of libgit2 until everything will be in order?

tkelman commented:
Sure, I don't see why not. Should be able to change `-DCMAKE_BUILD_TYPE` in `deps/Makefile`, clear out the buildbots and wait for a new windows nightly

tkelman commented:
A debug build of libgit2 should be in the Windows nightlies soon. Might be good to revert https://github.com/JuliaLang/julia/commit/71f08f30c915d3087c966daf66dd17c01b9092c5, or change it to `RelWithDebInfo`, within this PR.

When I do a local win32 build of this branch using a debug libgit2 build, it passes the pkg test. Though I do get a strange assertion failure on exit:
```
This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.
Assertion failed!

Program: D:\cygwin64\home\Tony\julia32\usr\bin\julia.exe
File: codegen.cpp, Line 737

Expression: !li->inInference
```

tkelman commented:
This is really strange. I can reproduce the segfault when I do a default Release build of libgit2, or a RelWithDebInfo build, but a Debug build passes. I wonder whose bug that is...

Tricky to get a backtrace at the moment because `julia-debug.exe` is broken.

wildart committed:
fixed 0.23 compatibility errors, added more commands to git-repl

wildart committed:
used new enums, changed location of GITHUB_REGEX, exported PkgError

wildart committed:
switch to release build for windows

wildart committed:
diff & rebase update to v0.23, error fixes

wildart committed:
added callbacks, updated push & publish

tkelman commented:
I came up with a C reproducer and reported the win32 issue upstream - https://github.com/libgit2/libgit2/issues/3342

We'll have to see what kind of patch we can come up with there.

tkelman commented:
Alright, I have a workaround in place. Homebrew has had the latest version for about a week, so rebase at your leisure. I doubt we'll get this in for 0.4, but it might be one of the first things we merge after branching.

IainNZ commented:
I guess it depends on how quick 0.5 really is, but if this indeed merged right after branching it could conceivably be backported for 0.4.1 right?
(I'm in favour of squeezing it in and shaking out any glitches during RC, but I'd understand not doing it)

tkelman commented:
Conceivably. We'll ship the libgit2 binaries in 0.4, there was a tentative idea to take them out if we're not using them yet but that was before this PR came together. This is the kind of thing that I'd really rather not rush, or hold up 0.4 over. But it will also end up similar to precompiled base or the new GC as a carrot that would convince more people to use master than really should be. Let's not pretend this is stable, or capable of getting there quickly, until we know it's stable.

tkelman commented on base/libgit2/types.jl:
```diff
@@ -0,0 +1,501 @@
+    perfdata_cb::Ptr{Void}
+    perfdata_payload::Ptr{Void}
+end
+CheckoutOptions(; checkout_strategy::Cuint = GitConst.CHECKOUT_SAFE,
```
All these default-kw immutables could still use a little macro I think to cut down on the repetition

wildart replied:
Maybe I'm waiting for default field values implementation, #10146, :wink:

tkelman commented on base/pkg/github.jl:
```diff
@@ -12,29 +12,30 @@ const AUTH_DATA = Dict{Any,Any}(
 
 function curl(url::AbstractString, opts::Cmd=``)
-    success(`curl --version`) || error("using the GitHub API requires having `curl` installed")
+    success(`curl --version`) || throw(PkgError("using the GitHub API requires having `curl` installed"))
```
Hm, we'll need to rewrite the curl parts of this for windows after we take out command-line git. Should be fairly straightforward to come up with replacements using powershell.

aviks replied:
Can't this use the `download` command in base, that is reasonably cross platform?
https://github.com/JuliaLang/julia/blob/master/base/interactiveutil.jl#L339-L369

tkelman replied:
No, this file needs to do a bunch more things like posting that `URLDownloadToFileW` can't do. Base's download is also kind of crippled and broken in a few ways like not following redirects correctly on Windows, ref https://github.com/JuliaLang/BinDeps.jl/issues/135. We could ship curl, but it would need some ca certificates and mozilla-nss and other nonsense to work correctly over https. `Invoke-RestMethod` in powershell is quite nice actually, having an object-oriented shell that understands json natively is pretty useful.

tkelman commented on base/methodshow.jl:
```diff
@@ -81,33 +81,34 @@ show(io::IO, mt::MethodTable) = show_method_table(io, mt)
-        end
-    end
-end
+#TODO: Convert to libgit2
```
how hard will this be?

wildart replied:
Should not be a problem. There are couple TODOs that are needed to be addressed after libgit2 wrapper is fully functional.

tkelman replied:
It isn't fully functional yet? What's still missing?

tkelman commented on base/libgit2.jl:
```diff
@@ -0,0 +1,492 @@
+                # head!(repo, hur, msg="--fastforward")
+
+                brn_ref_oid = Oid(brn_ref)
+                target!(head_ref, brn_ref_oid, msg="pkg.libgit2.megre!: fastforward $(name(brn_ref)) into $(name(head_ref))")
```
typo "megre" in msg

tkelman commented on base/pkg/entry.jl:
```diff
@@ -242,44 +288,53 @@ function free(pkgs)
+                LibGit2.fetch(repo)
+                LibGit2.checkout_head(repo)
+                LibGit2.branch!(repo, branch, track="refs/remotes/origin/$branch")
+                LibGit2.merge!(repo)
```
why is this merge instead of checkout branch?

wildart replied:
This should be a FF merge of a newly created branch with its tracking remote counterpart.

tkelman commented on base/pkg/entry.jl:
```diff
@@ -242,44 +288,53 @@ function free(pkgs)
+            end
         end
+        LibGit2.fetch(repo)
+        LibGit2.merge!(repo, fast_forward=true) || LibGit2.rebase!(repo, "origin/$branch")
```
`fast_forward=true` means this would fail any time the user has local commits in their metadata? the `pull --rebase` would be good to keep, so metadata pr's aren't full of merge commits.

hayd replied:
A fast forward merge is equivalent to a rebase (the RHS) if there are no local commits, so I think the LHS can be removed.

wildart replied:
True, but FF merge uses faster API then rebase.

tkelman replied:
I somehow completely missed the part after the `||`. Could use a few comments for non-obvious stuff like this.

wildart committed:
merge overhaul, added credential & fetchhead callbacks, small fixes & cleanup

wildart commented:
I guess I need to rebase for testing against appveyor. I'll do it later today.

tkelman commented:
We can temporarily add this branch to the appveyor whitelist if you're in a hurry, it's otherwise idle right now

wildart commented:
no, I'd rather rebase and see what will break.

wildart committed:
added extended repo open, ported Base.url

wildart committed:
fixed Pkg.pull_request, suppress backtrace PkgError & code cleanup

wildart commented:
@tkelman Green zone, at last.

tkelman commented on base/libgit2/repl.jl:
```diff
@@ -0,0 +1,349 @@
+
+List of git commands:
+- [x] add
+- [ ] branch
```
`- [ ] checkout` ?

wildart replied:
:dart:

tkelman commented on base/pkg/dir.jl:
```diff
@@ -26,21 +26,26 @@ function cd(f::Function, args...; kws...)
+        Base.cd(()->f(args...; kws...), dir)
+    catch err
+        if isa(err, PkgError)
+            print_with_color(:red, "PkgMgr Error: $(err.msg)")
```
I don't think we refer to it as `PkgMgr` anywhere else, I'd say either `PkgError: $(err.msg)` or spelled out as `Package manager error: $(err.msg)` would be clearer

wildart replied:
It was my initial version, but `Package manager error` was too long considering some error messages. I'll make it just 'ERROR`

tkelman commented on base/pkg/entry.jl:
```diff
@@ -333,38 +401,41 @@ function submit(pkg::AbstractString, commit::AbstractString="")
+        ahead_local == 0 && throw(PkgError("There are no METADATA changes to publish"))
+
+        # get changed files
+        for path in LibGit2.diff_files(repo, "origin/metadata-v2", LibGit2.GitConst.HEAD_FILE)
```
probably shouldn't be hard-coded

tkelman commented on base/pkg/github.jl:
```diff
@@ -12,29 +12,30 @@ const AUTH_DATA = Dict{Any,Any}(
+        throw(PkgError("""
         no GitHub user name configured; please configure it with:
 
             git config --global github.user USERNAME
```
This has to go through LibGit2 now, right?

wildart replied:
Sure, just need to add `config` command to Git REPL

tkelman commented on .travis.yml:
```diff
@@ -66,7 +66,7 @@ script:
     - cp /tmp/julia/lib/julia/sys.ji local.ji && /tmp/julia/bin/julia -J local.ji -e 'true' && /tmp/julia/bin/julia-debug -J local.ji -e 'true' && rm local.ji
     - /tmp/julia/bin/julia -e 'versioninfo()'
-    - export JULIA_CPU_CORES=4 && cd /tmp/julia/share/julia/test && /tmp/julia/bin/julia --check-bounds=yes runtests.jl all && /tmp/julia/bin/julia --check-bounds=yes runtests.jl pkg
+    - export JULIA_CPU_CORES=4 && cd /tmp/julia/share/julia/test && /tmp/julia/bin/julia --check-bounds=yes runtests.jl libgit2 repl && /tmp/julia/bin/julia --check-bounds=yes runtests.jl pkg
```
to be reverted

tkelman commented on appveyor.yml:
```diff
@@ -49,4 +49,4 @@ build_script:
   - usr\bin\julia -e "versioninfo()"
   - copy usr\lib\julia\sys.ji local.ji && usr\bin\julia -J local.ji -e "true" && del local.ji
-  - cd test && ..\usr\bin\julia runtests.jl all
+  - cd test && ..\usr\bin\julia runtests.jl libgit2 repl pkg
```
Not sure if `all pkg` will work correctly, but if it doesn't let's add a separate line that just runs `pkg`. It takes less than 2 minutes on appveyor and the queue has been fine lately so let's add it.

tkelman commented:
This is awesome. I guess people should give this branch a whirl and try to break it.

timholy commented:
I'm going to echo @IainNZ's comment at the top: heroic stuff, and I am really looking forward to it! Thanks in advance @wildart, and also @tkelman for being such a steadfast reviewer.

wildart committed:
switch GitRepo to GitRevWalker

wildart committed:
added to GitREPL: branch & tag

wildart committed:
added `commit` to Git REPL & fixed commands help info

wildart committed:
fix Git REPL `commit` output

wildart committed:
added `checkout` to Git REPL

wildart committed:
fixed remote branch name

wildart committed:
reomoved Git module and its tests

wildart committed:
fix error message & add all tests

wildart commented:
I say it's for testing. Do not forget to test Git REPL (press `>` to switch).

wildart committed:
fixed cache update, remove redundant `chomp` call from Git REPL `commit`,
added `fetch` and `push` to Git REPL

ScottPJones commented:
Wanted to test this, but ran into the conflicts.

wildart commented:
I'll wait for 0.4 release, and then continue to work on it. Meanwhile, I'll do updates for new `libgit2` versions and move all Git REPL code info separate branch.

wildart committed:
fixed: `tag` requirements creation, empty callbacks

tkelman commented:
Rebase when ready? I agree with splitting out the git repl stuff to separate concerns. You and Keno may have to fight over key bindings there.

wildart committed:
fixed tests

wildart committed:
removed Git REPL

wildart committed:
fixed bad rebase commit of `inbase`

tkelman commented on test/pkg.jl:
```diff
@@ -41,7 +41,6 @@ temp_pkg_dir() do
   Pkg.rm("Example")
   @test isempty(Pkg.installed())
   @test !isempty(Pkg.available("Example"))
-  @test_throws ErrorException Pkg.available("FakePackageDoesn'tExist")
```
why was this removed?

wildart replied:
Many `Pkg` calls are executed on a collection packages, so if something happened with the package the execution would continue on the rest. But, you are right. It was silly of me to remove the test because it can be addressed to submodule call, `Pkg.Entry.available`. I'll correct it.

wildart replied:
But the whole `Pkg` thing is totally botched with `cd` calls required for `git` commands. On the next iteration of the `Pkg` rewrite `cd` calls will be purged to make it work with absolute paths. And such tests will be easy to implement without directory changing shenanigans. See #12710.

tkelman replied:
This still needs to be put back.

ScottPJones commented:
Any ETA on merging this?

tlycken commented:
As a user who's switching between Ubuntu and Windows, the speed difference is excruciating... Is there anything I (as a Windows user) can do to help with this to speed up the merging?

I do realize that this won't make it into 0.4, but I'd still like to help this land as soon as possible, to make sure that it doesn't have to wait for 0.6.

tkelman commented:
Test and review this branch. Could build custom nightlies from this branch if it would help.

tlycken commented:
@tkelman Custom nightlies would be helpful. I haven't been able to set up a working build chain on my Win machine.

tkelman commented:
Check s3.amazonaws.com/julianightlies in an hour or so for binaries of 50142f0739

tlycken commented:
Wow, `Pkg` is blazingly fast now! :D I'll see if I can run through all `Pkg` functions at least once and see if I hit any problems. If there are any specific use cases or brittle areas that you want me to test more thoroughly, just let me know.

tkelman commented:
Bump. Can anyone else review this, or offer suggestions for ways in which it could be split up into smaller easier-to-merge separate PR's?

stevengj commented on base/libgit2/utils.jl:
```diff
@@ -0,0 +1,8 @@
+function version()
+    major, minor, patch = Cint[0], Cint[0], Cint[0]
+    ccall((:git_libgit2_version, :libgit2), Void,
+          (Ptr{Cint}, Ptr{Cint}, Ptr{Cint}), major, minor, patch)
```
The 0.4 style is to use `Ref`...

stevengj commented on base/libgit2/walker.jl:
```diff
@@ -0,0 +1,108 @@
+                  by::Cint = GitConst.SORT_NONE,
+                  rev::Bool=false,
+                  count::Int=0)
+    res = nothing
```
not type stable?

stevengj commented on base/libgit2/walker.jl:
```diff
@@ -0,0 +1,108 @@
+    while !done(walker, s)
+        val = f(s[1], repo)
+        if res == nothing
+            res = Array(typeof(val),0)
```
assumes `f` always returns the same type?

stevengj commented on base/libgit2/walker.jl:
```diff
@@ -0,0 +1,108 @@
+    while !done(walker, s) && val
+        val = f(s[1], repo)
+        _, s = next(walker, s)
+        c += 1
```
To be like `Base.count`, shouldn't this be `c += val == true`?

stevengj commented on base/pkg.jl:
```diff
@@ -2,15 +2,19 @@
+       build, free, pin, PkgError
 
-const DEFAULT_META = "git://github.com/JuliaLang/METADATA.jl"
+const DEFAULT_META = "https://github.com/JuliaLang/METADATA.jl"
```
Why the change?

tkelman replied:
maybe trying to address https://github.com/JuliaLang/julia/issues/7005 simultaneously - should probably be done as a separate pr though

wildart replied:
Yep, some things have been collected during the development.

stevengj commented on base/pkg/entry.jl:
```diff
@@ -46,22 +50,24 @@ function add(pkg::AbstractString, vers::VersionSet)
-                    outdated = Git.success(`diff --quiet origin/$branch`, dir="METADATA") ?
-                        (:no) : (:yes)
+        outdated = with(GitRepo, "METADATA") do repo
+            current_barnch = LibGit2.branch(repo)
```
barnch, not branch?

stevengj commented on base/pkg/entry.jl:
```diff
@@ -143,23 +162,43 @@ function status(io::IO, pkg::AbstractString, ver::VersionNumber, fix::Bool)
+    @printf io " - %-29s " pkg
+    @printf io "%-19s" msg
+    println(io)
+end
```
Can't all three lines be combined into a single `@printf io " - %-29s %-19s\n" pkg msg`?

wildart committed:
added: tree entry type, tree walker, blob lookup, some refinements
fixed: revwalker map type instability & count, typos

wildart committed:
Union() => Union{}

malmaud commented:
I tried using this branch and ran into some trouble:

```julia
julia> Pkg.update()
INFO: Updating METADATA...
INFO: Updating cache of Conda...
INFO: Updating cache of HDF5...
INFO: Updating cache of DualNumbers...
INFO: Updating CUDA...
INFO: Updating Atom...
INFO: Updating Shapefile...
INFO: Updating CUDArt...
INFO: Updating Gallium...
WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
INFO: Updating DIDebug...
WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
INFO: Updating IJulia...
INFO: Updating JITTools...
WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
INFO: Updating InControl...
WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
INFO: Updating MbedTLS...
INFO: Updating ObjFileBase...
INFO: Updating Gadfly...
INFO: Updating VT100...
WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
INFO: Updating Zlib...
INFO: Updating HttpServer...
INFO: Updating AbstractTrees...
WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
INFO: Updating HttpParser...
INFO: Updating Requests...
INFO: Updating DataFrames...
INFO: Updating MachO...
WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
INFO: Updating HttpCommon...
INFO: Updating TerminalUI...
WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
INFO: Updating Cxx...
INFO: Computing changes...
WARNING: DataFrames is fixed at 0.0.0- conflicting with requirement for Gadfly: [0.4.2,∞)
INFO: Upgrading Conda: v0.1.5 => v0.1.6
INFO: Upgrading DualNumbers: v0.1.3 => v0.1.4
INFO: Upgrading HDF5: v0.5.5 => v0.5.6
INFO: Building Conda
INFO: Recompiling stale cache file /Users/malmaud/.julia/lib/v0.5/Compat.ji for module Compat.
INFO: Recompiling stale cache file /Users/malmaud/.julia/lib/v0.5/JSON.ji for module JSON.
INFO: Recompiling stale cache file /Users/malmaud/.julia/lib/v0.5/DataStructures.ji for module DataStructures.
INFO: Recompiling stale cache file /Users/malmaud/.julia/lib/v0.5/URIParser.ji for module URIParser.
INFO: Recompiling stale cache file /Users/malmaud/.julia/lib/v0.5/SHA.ji for module SHA.
INFO: Building Homebrew
WARNING: could not import Pkg.Git into Homebrew
==========================================================[ ERROR: Homebrew ]===========================================================

LoadError: LoadError: UndefVarError: Git not defined
while loading /Users/malmaud/.julia/v0.5/Homebrew/src/Homebrew.jl, in expression starting on line 353
while loading /Users/malmaud/.julia/v0.4/Homebrew/deps/build.jl, in expression starting on line 1

========================================================================================================================================
INFO: Building Blosc
INFO: Building HDF5

============================================================[ BUILD ERRORS ]============================================================

WARNING: Homebrew had build errors.
```

jakebolewski commented:
The wrapper around command line Git in Base that Homebrew uses has been
removed, so this is not really a problem. The old git wrapper should
probably be put into package for package backward compatibility.

On Wed, Sep 23, 2015 at 9:31 PM, Jon Malmaud <notifications@github.com>
wrote:

> I tried using this branch and ran into some trouble:
>
> julia> Pkg.update()
> INFO: Updating METADATA...
> INFO: Updating cache of Conda...
> INFO: Updating cache of HDF5...
> INFO: Updating cache of DualNumbers...
> INFO: Updating CUDA...
> INFO: Updating Atom...
> INFO: Updating Shapefile...
> INFO: Updating CUDArt...
> INFO: Updating Gallium...
> WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
> INFO: Updating DIDebug...
> WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
> INFO: Updating IJulia...
> INFO: Updating JITTools...
> WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
> INFO: Updating InControl...
> WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
> INFO: Updating MbedTLS...
> INFO: Updating ObjFileBase...
> INFO: Updating Gadfly...
> INFO: Updating VT100...
> WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
> INFO: Updating Zlib...
> INFO: Updating HttpServer...
> INFO: Updating AbstractTrees...
> WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
> INFO: Updating HttpParser...
> INFO: Updating Requests...
> INFO: Updating DataFrames...
> INFO: Updating MachO...
> WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
> INFO: Updating HttpCommon...
> INFO: Updating TerminalUI...
> WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol
> INFO: Updating Cxx...
> INFO: Computing changes...
> WARNING: DataFrames is fixed at 0.0.0- conflicting with requirement for Gadfly: [0.4.2,∞)
> INFO: Upgrading Conda: v0.1.5 => v0.1.6
> INFO: Upgrading DualNumbers: v0.1.3 => v0.1.4
> INFO: Upgrading HDF5: v0.5.5 => v0.5.6
> INFO: Building Conda
> INFO: Recompiling stale cache file /Users/malmaud/.julia/lib/v0.5/Compat.ji for module Compat.
> INFO: Recompiling stale cache file /Users/malmaud/.julia/lib/v0.5/JSON.ji for module JSON.
> INFO: Recompiling stale cache file /Users/malmaud/.julia/lib/v0.5/DataStructures.ji for module DataStructures.
> INFO: Recompiling stale cache file /Users/malmaud/.julia/lib/v0.5/URIParser.ji for module URIParser.
> INFO: Recompiling stale cache file /Users/malmaud/.julia/lib/v0.5/SHA.ji for module SHA.
> INFO: Building Homebrew
> WARNING: could not import Pkg.Git into Homebrew==========================================================[ ERROR: Homebrew ]===========================================================
>
> LoadError: LoadError: UndefVarError: Git not definedwhile loading /Users/malmaud/.julia/v0.5/Homebrew/src/Homebrew.jl, in expression starting on line 353while loading /Users/malmaud/.julia/v0.4/Homebrew/deps/build.jl, in expression starting on line 1
> ========================================================================================================================================
> INFO: Building Blosc
> INFO: Building HDF5
> ============================================================[ BUILD ERRORS ]============================================================
>
> WARNING: Homebrew had build errors.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/pull/11196#issuecomment-142778416>.
>

tkelman commented:
We'll be taking command-line git out of the binaries on mac and windows, so that package would have to be responsible for installing git itself. Homebrew.jl and any other packages that were using git functionality through the Julia wrappers in base will need version-conditional code to use libgit2 instead.

jakebolewski commented:
Homebrew.jl uses the generic `Git.run` method, so it unclear whether the minimal amount of libgit wrapped for package functionality will be sufficient.

tkelman commented:
clone, config, fetch, and reset should all work?

jakebolewski commented:
Why don't we just merge this and clean up the pieces?

tkelman commented on base/libgit2.jl:
```diff
@@ -0,0 +1,500 @@
+        diff = diff_tree(repo, tree1, tree2)
+        for i in 1:count(diff)
+            delta = diff[i]
+            delta == nothing && break
```
comparisons to `nothing` should use `===` or `!==` throughout

tkelman commented:
> `WARNING: fetch: [Code:ERROR, Class:Net]: Unsupported URL protocol`

That is worrying and sounds like it could be a legitimate issue. Something may be wrong with how upstream is using the OSX system TLS library for https? Or with how we're using libgit2 here?

wildart commented:
@malmaud @tkelman It looks like `libgit2`library for OSX was built without `https` support.

>clone, config, fetch, and reset should all work?

yes, all routings are available in `LibGit2` module. Covering `git` calls in `Homebrew` should not be a problem.

tkelman commented on base/methodshow.jl:
```diff
@@ -89,7 +89,8 @@ function inbase(m::Module)
     end
 end
-fileurl(file) = let f = find_source_file(file); f === nothing ? "" : "file://"*f; end
+fileurl(file) = let f = find_source_file(file); f == nothing ? "" : "file://"*f; end
```
shouldn't be changing this?

tkelman commented:
@malmaud what osx version are you on? can you do `make -C deps distclean-libgit2 && make -C deps install-libgit2 > make.log 2>&1` and gist the log?

wildart committed:
fixed comparisons to `nothing`

malmaud commented:
https://gist.github.com/malmaud/4f5417f9c1eef27bd889

10.10.4

tkelman commented:
Great, thanks. So it built fine and said it found the relevant osx https libraries
```
-- Found Security /System/Library/Frameworks/Security.framework
-- Found CoreFoundation /System/Library/Frameworks/CoreFoundation.framework
```

Does `make -C deps check-libgit2` pass?

malmaud commented:
Yes it does.

wildart commented:
It looks like `ssh` protocol is not available in his build.
>-- checking for module 'libssh2'
>--   package 'libssh2' not found
>-- LIBSSH2 not found. Set CMAKE_PREFIX_PATH if it is installed outside of the default search path.

@malmaud What kind of protocol is used for any of packages that failed to `fetch`?

tkelman commented:
We never need to use ssh protocol do we?

wildart commented:
Usually, it is `git`+`ssh` combination, e.g. Github repos.

malmaud commented:
Looks like that unsupported URL editor is happening for repositories with SSH fetch URLs like `Fetch URL: git@github.com:Keno/Gallium.jl.git`

jakebolewski commented:
Yes, I think that  all the interactive prompting in https://github.com/JuliaLang/julia/pull/11196/files#diff-43fbc9e635c9a9f0da9707c61e66dea7R50 can go. Just error out and tell the user to set sane default locations.

aviks commented:
I agree all the prompting should go. But shouldn't this respect ssh_config for key location? Also, I doubt setting the key passphrase as an environment variable is a good idea... this should respect `ssh-agent`, on linux and osx.

jakebolewski commented:
After speaking with @tkelman I think the path forward here is to not expect libssh2 to be installed (Tony reports that this is difficult to install on Windows) and use https only for the Base Pkg operations.  Developers who want extra functionality (like ssh support) could install an add on package that extends the Base libgit2 functionality. 

For a smooth transition I think `base/pkg/pkg.jl` should be added back to this PR and the `Git` module re-instated until `Homebrew.jl` and other packages which rely on the cmdline git wrappers in Base case be transitioned over to the new API or a command line Git wrapper package is created.

StefanKarpinski commented:
Seems like a good plan.

wildart committed:
renamed module `GitConst` to `Consts`, added support of ssh-agent

wildart committed:
returned `Pkg.Git` module

wildart commented:
I returned `Pkg.Git` and made some corrections to credentials handling mechanism to allow ssh-agent support. BTW, `libgit2` does not offer much in SSH functionality - couple of functions that setup SSH credentials, and these functions will be disabled if library isn't compiled with `libssh` support. So, I do not see any way to support SSH unless we build `libgit2` properly or provide a way to customize all `ccall`'s to different build with SSH support. 

I like a mandatory support of `https` for all `Pkg` operations. As well as, a creation of the separate Julia Package SDK (`PkgDev` maybe) which would not only include SSH support but also will have all `Pkg` functions related to METADATA changes, so they could be remove form `Base` (`Pkg.update` will surely stay). This will allow to use bare repo for METADATA and keep it small (wrt to file number).

tkelman commented:
Yep, I agree with that plan now. Make base Pkg as small as it can be to just download and update packages, checkout and a few other things, and move most of the rest like generate, register, tag, publish out.

Put back the test `@test_throws ErrorException Pkg.available("FakePackageDoesn'tExist")` (or should it be `PkgError` now?) then I'm +1 to merge.

jakebolewski committed:
fix deprecation warning

jakebolewski committed:
add error branch test for package availablity

jakebolewski commented:
Thanks @wildart for being persistent and addressing all our points over these past months!

vtjnash commented:
this looks truly monumental. i'm excited!

jiahao commented:
:cake:

StefanKarpinski commented:
Oh sweet! This is really exciting.

tkelman commented:
Note that we'll probably have to split up any libgit2 tests that rely on internet connectivity, put them either in pkg or a new file that doesn't run by default (since distro buildbots don't have internet connectivity) but we manually add to travis and appveyor.

wildart commented:
What's happened? I was thinking about how to spit `Pkg` before merging and ... I guess it's for the best cause this PR has become so bloated I have a hard time tracking issues.

Thanks everybody for reviewing. @tkelman You've been a great help.

tlycken commented:
Great work, everyone involved! I'm really looking forward to using this! :)

KristofferC commented:
![tngnevc](https://cloud.githubusercontent.com/assets/1282691/10187727/afdf3114-675a-11e5-9770-416037908bcf.gif)

ScottPJones commented:
@wildart How will this be split, technically?  Will you put generic function definitions in Base, i.e.
`function name end`
and then via a "using PkgDev" add the actual methods for those (but they'd still be accessible as:
`Pkg.blahblahblah`?
I think that approach might be nicer than having to do `PkgDev.blahblahblah`, and remember which things are in Base, and which are extensions.

wildart commented:
Definitely `PkgDev`, no monkey business with `Base`.

ScottPJones commented:
Is it really monkey business to put generic definitions (which were added late in 0.4) for all the functions?
It seems like a nice technique that could be used in a lot of places where "advanced" stuff is planned to be moved out of Base in the future.

jakebolewski commented:
You can import `Base.Pkg` into `PkgDev and then re-export the methods, that way `PkgDev` becomes a strict superset in functionality (no cognitive burden required).

ScottPJones commented:
But then you still have both `Pkg` and `PkgDev`, and only some methods are available in `Pkg`.
That seems like a bother to me.

wildart commented:
The whole idea of `PkgDev` is separating a development functionality from `Pkg` which should target user experience only.

hayd commented:
Also #5155.

tkelman commented:
@wildart did you ever get a chance to structure the Git REPL mode as a package? A few people have expressed interest in using it, would be cool to have available to try out somewhere.
